synthesizer/src/vm/finalize.rs

```diff
@@ -263,10 +263,10 @@ impl<N: Network, C: ConsensusStorage<N>> VM<N, C> {
 
             /* Perform the atomic finalize over the transactions. */
 
-            // Acquire the write lock on the process.
+            // Acquire the read lock on the process.
             // Note: Due to the highly-sensitive nature of processing all `finalize` calls,
-            // we choose to acquire the write lock for the entire duration of this atomic batch.
-            let process = self.process.write();
+            // we choose to acquire the read lock for the entire duration of this atomic batch.
+            let process = self.process.read();
```

Comment:
The point of the write lock here is so that there is NO concurrency on finalize scope state. Because transaction verification is highly dependent on state that can be changed between transactions we would much rather lean on safety guarantees.

Imagine transaction A and B where they both spend 10 credits from the same mapping. If they are read concurrently, both are considered valid, which should not be the case.

 In addition the `atomic_finalize` function below also uses this same `let process = self.process.write();` pattern; this breaks the consistency.

At the very least there should be some testing guarantee the safety of this change, but **I do not believe this is safe**.

Reply:
Since these strict requirements go against the assumption from the description, my recommendation is to close this PR and I'll soon propose one extending the related comments/docs so that it is clear in the future :+1:.

Reply:
+1 on closing this PR, while I think this PR is safe, it is not putting any guardrails in place to ensure future developments remain safe. Some comments to grow our (mostly my own) understanding:

> In addition the atomic_finalize function below also uses this same let process = self.process.write(); pattern; this breaks the consistency.

Indeed, but in contrast to finalize (which needs write access to *finalize*), speculate conceptually only needs read access.

> Imagine transaction A and B where they both spend 10 credits from the same mapping. If they are read concurrently, both are considered valid, which should not be the case.

`self.atomic_lock` prevents this. Moreover, what is critical: `Process` doesn't contain data related to transaction verification. So maybe the comment could be adjusted to something like: "There are no guardrails in place to ensure Process will never contain data relevant for transmission verification".

Maybe a targeted unit test or some smart types could be sufficient for that, but is likely too much effort for now.

Reply:
Agreed, `self.atomic_lock` does indeed prevent the double-spend issue.

I still think that any changes related to the finalize scope and state management requires more scrutiny and additional testing.